### PR TITLE
[CH-186] support `RangePartitioning`

### PR DESCRIPTION
**Pull request by lgbo-ustc**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
see issue #186 

We implement `RangePartitionNativeSplitter` in this pr which support running range partitioning in native backend.
This pr is mainly for improving the `order by` clause.

Some benchmark test by running `order by`.
|  data type   |  rows  | eclapsed time(s) by gluten  |  eclapsed time(s) by spark  | 
|  ----  | ----  | ----  |  ----  | 
| long | 69498916 | 6.9 | 8.5 |
| long | 208496748 | 13.5 | 22.8 |
| string | 69498916 | 10.5 | 26.5 |
| string | 208496748 | 23.9 | 71.3 |

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by kyligence-git** [comment]
Can one of the admins verify this patch?

**Comment by lgbo-ustc** [comment]
rely on https://github.com/oap-project/gluten/pull/524

**Comment by lgbo-ustc** [comment]
test this please with 524

**Comment by lgbo-ustc** [comment]
test this please with 524

**Comment by lgbo-ustc** [comment]
test this please with 524

**Comment by lgbo-ustc** [comment]
test this please with 524

File: utils/local-engine/Shuffle/NativeSplitter.cpp [hunk 1]
```diff
@@ -176,4 +233,225 @@ void RoundRobinNativeSplitter::computePartitionId(Block & block)
...
+                    col->insert(val);
+                }
+                else
+                {
+                    throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Unsupported data type: {}", type_info.inner_type->getName());
+                }
+            }
+        }
+        auto data_type = type_info.inner_type;
+        if (type_info.is_nullable)
```
Comment (zhanglistar) [inline utils/local-engine/Shuffle/NativeSplitter.cpp:347]: can merge with line 289

File: utils/local-engine/Shuffle/NativeSplitter.cpp [hunk 2]
```diff
@@ -176,4 +233,225 @@ void RoundRobinNativeSplitter::computePartitionId(Block & block)
...
+}
+
+int RangePartitionNativeSplitter::compareRow(
+    const DB::Columns & columns,
+    const std::vector<size_t> & required_columns,
+    size_t row,
+    const DB::Columns & bound_columns,
+    size_t bound_row)
+{
+    int result = 0;
```
Comment (zhanglistar) [inline utils/local-engine/Shuffle/NativeSplitter.cpp:392]: delete

File: utils/local-engine/Shuffle/NativeSplitter.cpp [hunk 3]
```diff
@@ -176,4 +233,225 @@ void RoundRobinNativeSplitter::computePartitionId(Block & block)
...
+    for(size_t i = 0, n = required_columns.size(); i < n; ++i)
+    {
+        auto lpos = required_columns[i];
+        auto rpos = i;
+        auto res = columns[lpos]->compareAt(row, bound_row, *bound_columns[rpos], sort_descriptions[i].nulls_direction)
+            * sort_descriptions[i].direction;
+        if (res != 0)
+        {
+            result = res;
```
Comment (zhanglistar) [inline utils/local-engine/Shuffle/NativeSplitter.cpp:401]: return res;

File: utils/local-engine/Shuffle/NativeSplitter.cpp [hunk 4]
```diff
@@ -176,4 +233,225 @@ void RoundRobinNativeSplitter::computePartitionId(Block & block)
...
+            break;
+        }
+    }
+    return result;
```
Comment (zhanglistar) [inline utils/local-engine/Shuffle/NativeSplitter.cpp:405]: return 0;

File: utils/local-engine/Shuffle/NativeSplitter.cpp [hunk 5]
```diff
@@ -176,4 +233,225 @@ void RoundRobinNativeSplitter::computePartitionId(Block & block)
...
+        {2, {1, 1}},
+        {3, {-1, 1}},
+        {4, {-1, -1}}
+};
+
+RangePartitionNativeSplitter::RangePartitionNativeSplitter(NativeSplitter::Options options_, jobject input)
+    : NativeSplitter(options_, input)
+{
+    Poco::JSON::Parser parser;
+    auto info = parser.parse(options_.exprs_buffer).extract<Poco::JSON::Object::Ptr>();
```
[the remark this answers is not in the document]
Comment (lgbo-ustc) [reply utils/local-engine/Shuffle/NativeSplitter.cpp:248]: Just throw the exception and corrupt the process

**Comment by lgbo-ustc** [comment]
test this please with 524

**Comment by zhanglistar** [comment]
As comment, there is a optimization to do: 
1. clickhouse backend consume data from JVM, but not sort the block at once, instead, we read the data and do nothing until memory limit is reached or all data is consumed
2. sort the data, here if memory limit is reached first, we then flush the sorted data to disk, and do 1; else return the output data stream.
3. do merge sort on the flushed data, and then return the output data stream.

**Comment by liuneng1994** [comment]
test this please with 524

**Comment by lgbo-ustc** [comment]
test this please with 524

**Review by liuneng1994 — APPROVED** [review]
LGTM

**Pull request merged**